chain-abci/src/app/rewards.rs

```diff
@@ -0,0 +1,193 @@
+use std::f64::consts::E;
+
+use chain_core::init::{coin::Coin, params::NetworkParameters};
+use chain_core::state::account::StakedStateAddress;
+use chain_core::state::tendermint::TendermintValidatorAddress;
+use chain_core::tx::fee::Milli;
+
+use crate::app::{update_account, ChainNodeApp};
+use crate::enclave_bridge::EnclaveProxy;
+use crate::storage::tx::get_account;
+
+fn milli_to_f64(milli: Milli) -> f64 {
+    milli.as_millis() as f64 / 1000_f64
+}
+
+/// FIXME more stable arithmetic
+pub fn monetary_expansion(
+    tau: Milli,
+    total_staking: Coin,
+    minted: Coin,
+    params: &NetworkParameters,
+) -> Coin {
+    let cap = params.get_rewards_monetary_expansion_cap();
+    let r0 = milli_to_f64(params.get_rewards_monetary_expansion_r0());
+    let tau = milli_to_f64(tau);
+    let total_staking = u64::from(total_staking) as f64 / 1_0000_0000_f64;
+    let amount = total_staking * r0 * E.powf(-total_staking / tau);
+    min(
+        (cap - minted).unwrap_or_default(),
+        Coin::new(amount as u64).unwrap(),
+    )
+}
+
+type RewardsDistribution = Vec<(StakedStateAddress, Coin)>;
+impl<T: EnclaveProxy> ChainNodeApp<T> {
+    /// Distribute rewards pool
+    pub fn rewards_try_distribute(&mut self) -> Option<(RewardsDistribution, Coin)> {
+        let state = self.last_state.as_mut().unwrap();
+
+        if state.block_time < state.genesis_time
+            || state.block_time < state.rewards_pool.last_distribution_time
+        {
+            // FIXME use global overflow/underflow check.
+            panic!("invalid block time");
+        }
+
+        if state.block_time - state.rewards_pool.last_distribution_time
+            < state.network_params.get_rewards_distribution_period() as u64
+        {
+            return None;
+        }
+        self.rewards_pool_updated = true;
+
+        let mut total_staking = Coin::zero();
+        for (addr, _) in self.validator_voting_power.iter() {
+            let account = get_account(addr, &state.last_account_root_hash, &self.accounts)
+                .expect("io error or validator account not exists");
+            total_staking = (total_staking + account.bonded).expect("coin overflow");
+        }
+
+        let minted = monetary_expansion(
+            state.rewards_pool.tau,
+            total_staking,
+            state.rewards_pool.minted,
+            &state.network_params,
+        );
+        // tau decay
+        state.rewards_pool.tau =
+            state.rewards_pool.tau * state.network_params.get_rewards_monetary_expansion_decay();
+        let total_rewards = (state.rewards_pool.remaining + minted).unwrap();
+        state.rewards_pool.minted = (state.rewards_pool.minted + minted).unwrap();
+
+        // state.rewards_pool.remaining
+        let total_blocks = state.proposer_stats.iter().map(|(_, count)| count).sum();
+
+        let share = (total_rewards / total_blocks).unwrap();
+        state.rewards_pool.remaining = (total_rewards % total_blocks).unwrap();
```

Comment:
I'm not sure what this is doing -- the original meaning of "remaining" was the total amount that's available to distribute from (i.e. some amount inserted at genesis + accumulation from fees and punishments)... there should be a fixed supply.


I'll check the spec in chain-docs

Reply:
Current meaning of `remaining` is the rewards pool to be distributed in the current period, the fixed supply is checked in `monetary_expansion`, and all the new minted coins is recorded in the field `minted`.

Reply:
ok, the original name meant remaining as in "remaining to distribute", which accumulated all funds to distribute -- would be good to update the comment: https://github.com/crypto-com/chain/pull/648/files#diff-f04e1c94e8043c78235a0960fa4fa25bR22

And perhaps, if now, the emission is split between a portion of `monetary_expansion_cap` and this extra `remaining` bit, it may be better to rename it to reflect the new meaning -- e.g. `period_bonus` or how to call it @lezzokafka ?

Reply:
Something like `current_rewards`?

Reply:
Renaming and comment change are done and force pushed.